### PR TITLE
Added Getting Started section in the README

### PR DESCRIPTION
**Pull request by lunchboxav**
This PR addresses #384. 

## Motivation

I'm adding a Getting Started section in the README, which includes code and explanation. Hopefully, this will ease new users of tracing to get up to speed. The current doc is amazing, but adding an explicit Getting Started section might also be useful. 

## Solution

a Getting Started section with code example and a brief explanation of what the code does.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,12 +34,93 @@ Application-level tracing for Rust.
 structured, event-based diagnostic information. `tracing` is maintained by the
 Tokio project, but does _not_ require the `tokio` runtime to be used.
 
+## Getting Started
+
+To get started, configure `Cargo.toml` to include `tracing` and `tracing-subscriber` as dependencies. `tracing` is the core library, while `tracing-subscriber` is the part that implements `Subscriber`, which will record the trace. Take a look at the [Project layout](#Project layout) section.
```
Comment (hawkw) [inline README.md]: How about:
```suggestion
To get started, add dependencies on `tracing` and `tracing-subscriber` to your `Cargo.toml`. The `tracing` crate provides the _instrumentation_ APIs for emitting traces, while `tracing-subscriber` provides implementations of the `Subscriber` trait, which will record the trace. For more information on the crates that make up the `tracing` project, take a look at the [Project layout](#Project layout) section.
```

File: README.md [hunk 2]
```diff
@@ -34,12 +34,93 @@ Application-level tracing for Rust.
...
+
+```
+[dependencies]
+tracing = "0.1.5"
+tracing-subscriber = "0.1.5"
+```
+
+Next, in our `main.rs` file, we will start by adding the libraries
+
+```
```
Comment (hawkw) [inline README.md]: Add 
```suggestion
```rust
```
to get Rust syntax highlighting here.

File: README.md [hunk 3]
```diff
@@ -34,12 +34,93 @@ Application-level tracing for Rust.
 structured, event-based diagnostic information. `tracing` is maintained by the
 Tokio project, but does _not_ require the `tokio` runtime to be used.
 
+## Getting Started
+
+To get started, configure `Cargo.toml` to include `tracing` and `tracing-subscriber` as dependencies. `tracing` is the core library, while `tracing-subscriber` is the part that implements `Subscriber`, which will record the trace. Take a look at the [Project layout](#Project layout) section.
+
+```
```
Comment (hawkw) [inline README.md]: Add 
```suggestion
```toml
```
to get toml syntax highlighting.

File: README.md [hunk 4]
```diff
@@ -34,12 +34,93 @@ Application-level tracing for Rust.
...
+Next, in our `main.rs` file, we will start by adding the libraries
+
+```
+use tracing::{debug, info, instrument};
+use tracing_subscriber::fmt;
+```
+
+For this simple example, we will create a function that prints sequence of numbers and we will instrument that cod to enable tracing. Tracing is enabled by initiating a span and the `#[instrument]` attribute, implicitly create a span for that function.
+
+```
```
Comment (hawkw) [inline README.md]:
```suggestion
```rust
```

File: README.md [hunk 5]
```diff
@@ -34,12 +34,93 @@ Application-level tracing for Rust.
...
+// using instrument to setup a span and skip to exclude args
+#[instrument(skip(_too_long_string))]
+fn count(nums: i32, _too_long_string: &str) -> Vec<i32> {
```
Comment (hawkw) [inline README.md]: What do you think about starting with a basic example of the use of `instrument`, like:
```rust
#[instrument]
fn count(nums: i32) -> Vec<i32>
```
and _then_ introducing the `skip` argument after showing the basic usage first?

File: README.md [hunk 6]
```diff
@@ -34,12 +34,93 @@ Application-level tracing for Rust.
...
+
+To get started, configure `Cargo.toml` to include `tracing` and `tracing-subscriber` as dependencies. `tracing` is the core library, while `tracing-subscriber` is the part that implements `Subscriber`, which will record the trace. Take a look at the [Project layout](#Project layout) section.
+
+```
+[dependencies]
+tracing = "0.1.5"
+tracing-subscriber = "0.1.5"
+```
+
+Next, in our `main.rs` file, we will start by adding the libraries
```
Comment (hawkw) [inline README.md]: How about:
```suggestion
Next, we will import those crates in our `main.rs` file:
```

File: README.md [hunk 7]
```diff
@@ -34,12 +34,93 @@ Application-level tracing for Rust.
...
+
+```
+use tracing::{debug, info, instrument};
+use tracing_subscriber::fmt;
+```
+
+For this simple example, we will create a function that prints sequence of numbers and we will instrument that cod to enable tracing. Tracing is enabled by initiating a span and the `#[instrument]` attribute, implicitly create a span for that function.
```
Comment (hawkw) [inline README.md]: Typo:
```suggestion
For this simple example, we will create a function that prints sequence of numbers and we will instrument that code to enable tracing. Tracing is enabled by initiating a span and the `#[instrument]` attribute, implicitly create a span for that function.
```

Comment (hawkw) [reply README.md]: How about
```suggestion
For this simple example, we will create a function that prints sequence of numbers and we will instrument that cod to emit traces. Tracing is enabled by initiating a span and the `#[instrument]` attribute, implicitly create a span for that function.
```

File: README.md [hunk 8]
```diff
@@ -34,12 +34,93 @@ Application-level tracing for Rust.
...
+        // logging n in structured form
+        debug!(n, "Next Number");
+        seq.push(n);
+    }
+
+    seq
+}
+```
+
+Then, in our `main()` function, we can create a subscriber, that records the trace. The subscriber employs an environment filter that we can utilize to filter the level of tracing message.
```
Comment (hawkw) [inline README.md]: How about:
```suggestion
Once we have added instrumentation to our code, we will need to set up a `Subscriber` to record traces. We'll do this in our `main()` function. The subscriber employs an environment filter that we can utilize to filter the level of tracing message.
```

Comment (hawkw) [reply README.md]: Also, we should probably link to the `Subscriber` trait's documentation at https://docs.rs/tracing/0.1.10/tracing/trait.Subscriber.html and possibly also to the docs on setting a subscriber at https://docs.rs/tracing/0.1.10/tracing/dispatcher/index.html#using-the-trace-dispatcher

File: README.md [hunk 9]
```diff
@@ -66,24 +147,24 @@ state, and are less stable than the `tracing` and `tracing-core` crates.
 
 The crates included as part of Tracing are:
 
-* [`tracing-futures`]: Utilities for instrumenting `futures`.
+- [`tracing-futures`]: Utilities for instrumenting `futures`.
```
Comment (hawkw) [inline README.md]: The change in markdown here seems unrelated to the rest of the change...was this suggested by `markdownlint` or something?

File: README.md [hunk 10]
```diff
@@ -94,13 +175,10 @@ The crates included as part of Tracing are:
 [`tracing-serde`]: tracing-serde
 [`tracing-subscriber`]: tracing-subscriber
 [`tracing-tower`]: tracing-tower
-
```
Comment (hawkw) [inline README.md]: More unrelated changes.

File: README.md [hunk 11]
```diff
@@ -34,12 +34,93 @@ Application-level tracing for Rust.
...
+
+    seq
+}
+```
+
+Then, in our `main()` function, we can create a subscriber, that records the trace. The subscriber employs an environment filter that we can utilize to filter the level of tracing message.
+
+```
+fn main() {
+  // this creates subscriber with an environment filter
```
Comment (hawkw) [inline README.md]: Personally, I'd prefer to start by introducing how to create and set a subscriber with the default filter
```rust
let subscriber = fmt::Subscriber::default();
tracing::subscriber::with_default(subscriber, || {
    // ...
});
```
and then introduce customizing the filter _after_ we demonstrate the default...

File: README.md [hunk 12]
```diff
@@ -34,12 +34,93 @@ Application-level tracing for Rust.
...
+        seq.push(n);
+    }
+
+    seq
+}
+```
+
+Then, in our `main()` function, we can create a subscriber, that records the trace. The subscriber employs an environment filter that we can utilize to filter the level of tracing message.
+
+```
```
Comment (hawkw) [inline README.md]:
```suggestion
```rust
```

File: README.md [hunk 13]
```diff
@@ -34,12 +34,93 @@ Application-level tracing for Rust.
...
+  let filter = tracing_subscriber::EnvFilter::try_from_default_env()
+      .unwrap_or_else(|_| DEFAULT_FILTER.into());
+  let subscriber = fmt::Subscriber::builder().with_env_filter(filter).finish();
+  ...
+}
+```
+
+Finally, we can run the function that prints out the tracing message while it's being run
+
+```
```
Comment (hawkw) [inline README.md]:
```suggestion
```rust
```

File: README.md [hunk 14]
```diff
@@ -34,12 +34,93 @@ Application-level tracing for Rust.
...
+Then, in our `main()` function, we can create a subscriber, that records the trace. The subscriber employs an environment filter that we can utilize to filter the level of tracing message.
+
+```
+fn main() {
+  // this creates subscriber with an environment filter
+  const DEFAULT_FILTER: &str = concat!(module_path!(), "=", "info");
+  let filter = tracing_subscriber::EnvFilter::try_from_default_env()
+      .unwrap_or_else(|_| DEFAULT_FILTER.into());
+  let subscriber = fmt::Subscriber::builder().with_env_filter(filter).finish();
+  ...
```
Comment (hawkw) [inline README.md]: it would be nice if this was valid rust code that actually compiles...we could make the "..." a comment so it's not a syntax error:
```suggestion
  // ...
```

File: README.md [hunk 15]
```diff
@@ -34,12 +34,93 @@ Application-level tracing for Rust.
...
+}
+```
+
+Finally, we can run the function that prints out the tracing message while it's being run
+
+```
+fn main() {
+  ...
```
Comment (hawkw) [inline README.md]: again, consider:
```suggestion
  // ...
```

**Comment by lunchboxav** [comment]
Thank you very much for the thorough comments @hawkw. I'll improve the PR :)

File: README.md (hunk 5 above)
Comment (lunchboxav) [reply README.md]: I'm, on board with this. I was thinking that I could introduce an argument to the instrument and explain it later down the road. But, yes, let's go with basic usage first.

File: README.md (hunk 7 above)
Comment (lunchboxav) [reply README.md]: agree 👍

File: README.md (hunk 9 above)
Comment (lunchboxav) [reply README.md]: ah, my bad, I think that's my lint doing.. will undo the changes

**Comment by hawkw** [comment]
hi @lunchboxav, now that the "batteries-included" [`init`](https://docs.rs/tracing-subscriber/0.1.6/tracing_subscriber/fmt/fn.init.html) and [`try_init`](https://docs.rs/tracing-subscriber/0.1.6/tracing_subscriber/fmt/fn.try_init.html) functions in `tracing-subscriber::fmt` have been released in version 0.1.6, can we use them in the introductory docs when applicable? should be more ergonomic for simple tasks

**Comment by lunchboxav** [comment]
sure thing @hawkw . I'll play around with `init` and `try-init` and I'll update the doc. Should be good for setting up subscriber for getting started.

**Comment by lunchboxav** [comment]
PTAL @hawkw. I'm torn between keeping example of using subscriber's `default` or just remove it since we have `init`. I'm keeping both for now, but I'm open to the idea of just use `init` as a preferred way to get started with `tracing`.

**Comment by lunchboxav** [comment]
Should I close this PR as well @hawkw? Considering that the issue this PR wanted to solve has already solved.

**Comment by hawkw** [comment]
@lunchboxav hmm... sorry about that! i'll have to take a closer look at the examples in this PR vs the ones added in #496, as we may still want to add some of the examples in this branch as well!

**Comment by lunchboxav** [comment]
That’s completely fine @hawkw. I haven’t had the time too keep up with the recent movement in this repo, so I may missed some significant bits. Just let me know what I can help in this PR :)
